projects/07/VMTranslator.py

```diff
@@ -0,0 +1,346 @@
+        elif segment in [POINTER, TEMP]:
+            self.offset_literal(segment, index)
+            self.write_line('A=D')
+            self.write_line('D=M')
+        else:
+            #static
+            static_label=self.generate_label(index)
+            self.write_line('@{label}'.format(label=static_label))
+            self.write_line('D=M')
+        self.write_line('@SP')
+        self.write_line('A=M')
+        self.write_line('M=D')
+        self.write_increment_sp()
+
+    def write_pop(self, segment, index):
+        if segment in [LOCAL, ARGUMENT, THIS, THAT]:
+            self.offset_segment(segment, index)
+        elif segment in [POINTER, TEMP]:
+            self.offset_literal(segment, index)
+        else:
+            #static
+            static_label = self.generate_label(index)
+            self.write_line('@{label}'.format(label=static_label))
+            self.write_line('D=A')
+        self.write_line('@SP')
+        self.write_line('A=M')
+        self.write_line('M=D')
+        self.write_decrement_sp()
+        self.write_line('D=M')
+        self.write_line('@SP')
+        self.write_line('A=M+1')
+        self.write_line('A=M')
+        self.write_line('M=D')
+
+    def offset_literal(self, segment, index):
+        self.write_line('@{seg}'.format(seg=SEGMENTS[segment]))
+        self.write_line('D=A')
+        self.write_line('@{index}'.format(index=index))
+        self.write_line('D=D+A')
+
+    def offset_segment(self, segment, index):
+        self.write_line('@{seg}'.format(seg=SEGMENTS[segment]))
+        self.write_line('D=M')
+        self.write_line('@{index}'.format(index=index))
+        self.write_line('D=D+A')
+
+    def write_line(self, line):
+        self.file.write(line)
+        self.file.write(NEWLINE)
+
+    def close(self):
+        '''
+        Closes the output file
+        '''
+        self.file.close()
+
+def translate(parser, writer):
+    while parser.has_more_commands():
+        parser.advance()
+        if parser.command_type() == CmdType.C_ARITHMETIC:
+            writer.write_arithmetic(parser.arg1())
+        if parser.command_type() == CmdType.C_POP or parser.command_type() == CmdType.C_PUSH:
+            writer.write_push_pop(parser.command_type(), parser.arg1(), parser.arg2())
+
+def translate_files(files, write_name):
+    writer = CodeWriter(write_name)
+    for file_name in files:
+        print('parsing {file}'.format(file=file_name))
+        parser = Parser(file_name)
+        writer.set_file_name(file_name)
+        translate(parser, writer)
+    writer.close()
+
+def get_write_path(dir, name):
+    write_path = Path(dir)
+    return write_path.joinpath(name + ASSEMBLY_EXTENSION)
+
+def main():
+    path = Path(sys.argv[1])
+    files = []
+    write_name = ''
+    if path.is_file():
+        files.append(path)
+        name, _ = path.name.split('.')
+        write_name = get_write_path(path.parent, name)
+    else:
+        for file_name in path.iterdir():
```

Comment:
Tested when I wrote the skeleton of this module and when using the folder name, but I'm worried that multifile vm code is gonna fail

Reply:
Project 8 has sys and main vms for the test files. It will be a trial by fire